ethylene-core/src/main/java/com/github/steanky/ethylene/core/processor/FileConfigLoader.java

```diff
@@ -0,0 +1,124 @@
+    private ConfigBridge bridge;
+
+    /**
+     * General implementation of {@link ConfigBridge} which calls {@link FileConfigLoader#codecForPath(Path)} to obtain
+     * a codec. If this call results in a null value (indicating a codec could not be found), exceptions may be thrown
+     * when attempting to retrieve the result of a {@link CompletableFuture} returned by
+     * {@link ValidatingConfigBridge#read()} or {@link ValidatingConfigBridge#write(ConfigElement)}.
+      */
+    protected abstract class ValidatingConfigBridge implements ConfigBridge {
+        /**
+         * The {@link ConfigCodec} instance used to decode file data, or null if there is no codec associated with the
+         * file
+         */
+        protected final ConfigCodec codec = codecForPath(path);
+
+        @Contract("null, _ -> fail")
+        private static void validateCodec(ConfigCodec codec, Path path) throws IOException {
+            if(codec == null) {
+                throw new IOException("No codec found for path " + path);
+            }
+        }
+
+        @Override
+        public @NotNull CompletableFuture<ConfigElement> read() {
+            return makeFuture(() -> {
+                validateCodec(codec, path);
+                codec.decode(Files.newInputStream(path));
+                return null;
+            });
+        }
+
+        @Override
+        public @NotNull CompletableFuture<Void> write(@NotNull ConfigElement element) {
+            return makeFuture(() -> {
+                validateCodec(codec, path);
+                codec.encode(element, Files.newOutputStream(path));
+                return null;
+            });
+        }
+
+        /**
+         * Produces a {@link CompletableFuture} object from a {@link Callable}. The future may or may not be already
+         * completed.
+         * @param callable the callable to convert
+         * @param <TReturn> the type of data object
+         * @return a {@link CompletableFuture} object which may represent a synchronous or asynchronous call
+         */
+        protected abstract <TReturn> @NotNull CompletableFuture<TReturn> makeFuture(
+                @NotNull Callable<TReturn> callable);
+    }
+
+    /**
+     * Constructs a new FileConfigLoader instance from the given {@link ConfigProcessor}, data object, and {@link Path}.
+     * @param processor the processor used to marshal data
+     * @param defaultData the default data object
+     * @param path the path to read data from and write defaults to
+     */
+    public FileConfigLoader(@NotNull ConfigProcessor<TData> processor,
+                            @NotNull TData defaultData,
+                            @NotNull Path path) {
+        super(processor, defaultData);
+        this.path = Objects.requireNonNull(path);
+    }
+
+    @Override
+    protected @NotNull ConfigBridge getBridge() {
+        if(bridge != null) {
+            return bridge;
+        }
+
+        return bridge = new ValidatingConfigBridge() {
```

Comment:
Why lazy init

Reply:
because this code (very indirectly) ends up referencing codecForPath which is an instance method that shouldn't really be called in the constructor

Reply:
maybe codecForPath shouldn't even be part of this class then?

Reply:
sure, can just make it a Function<Path, ConfigCodec>